tests/integ/scala/test_async_job_suite.py

```diff
@@ -0,0 +1,267 @@
+    res = target.delete(target["a"] == 1, block=False)
+    assert res.result() == DeleteResult(rows_deleted=2)
+    Utils.check_answer(
+        target, [Row(A=2, B=1), Row(A=2, B=2), Row(A=3, B=1), Row(A=3, B=2)]
+    )
+
+    # update operation
+    target_df = session.create_dataframe(
+        [(1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2)], schema=["a", "b"]
+    )
+    target_df.write.save_as_table("my_table", mode="overwrite", table_type="temporary")
+    target = session.table("my_table")
+    res = target.update({"b": 0, "a": target.a + target.b}, block=False)
+    assert res.result() == UpdateResult(rows_updated=6, multi_joined_rows_updated=0)
+    Utils.check_answer(
+        target,
+        [
+            Row(A=2, B=0),
+            Row(A=3, B=0),
+            Row(A=3, B=0),
+            Row(A=4, B=0),
+            Row(A=4, B=0),
+            Row(A=5, B=0),
+        ],
+    )
+
+
+def test_async_save_as_table(session):
+    df = session.create_dataframe(
+        [[float("nan"), 3, 5], [2.0, -4, 7], [3.0, 5, 6], [4.0, 6, 8]],
+        schema=["a", "b", "c"],
+    )
+    table_name = random_name_for_temp_object(TempObjectType.TABLE)
+    async_job = df.write.save_as_table(table_name, create_temp_table=True, block=False)
+    assert async_job.result() is None
+    table_df = session.table(table_name)
+    Utils.check_answer(table_df, df)
+
+
+def test_async_copy_into_location(session):
+    remote_location = f"{session.get_session_stage()}/{random_name_for_temp_object(TempObjectType.TABLE)}.csv"
+    df = session.create_dataframe(
+        [[float("nan"), 3, 5], [2.0, -4, 7], [3.0, 5, 6], [4.0, 6, 8]],
+        schema=["a", "b", "c"],
+    )
+    test_schema = StructType(
+        [
+            StructField("a", DoubleType()),
+            StructField("b", IntegerType()),
+            StructField("c", IntegerType()),
+        ]
+    )
+    # check if copy is successful
+    async_job = df.write.copy_into_location(remote_location, block=False)
+    res = async_job.result()
+    Utils.check_answer(res, [Row(rows_unloaded=4, input_bytes=27, output_bytes=47)])
+
+    # check the content of copied table
+    res = session.read.schema(test_schema).csv(remote_location)
+    Utils.check_answer(res, df)
+
+
+def test_multiple_queries(session, resources_path):
+    user_schema = StructType(
+        [
+            StructField("a", IntegerType()),
+            StructField("b", StringType()),
+            StructField("c", DoubleType()),
+        ]
+    )
+    test_files = TestFiles(resources_path)
+    Utils.create_stage(session, tmp_stage_name1, is_temporary=True)
+    Utils.upload_to_stage(
+        session, "@" + tmp_stage_name1, test_files.test_file_csv, compress=False
+    )
+    df = session.read.schema(user_schema).csv(f"@{tmp_stage_name1}/{test_file_csv}")
+    assert len(df.queries) > 1
+    res = df.collect_nowait()
+    Utils.check_answer(res.result(), df.collect())
+
+
+def test_async_is_running_and_cancel(session):
+    pass
```

Comment:
You need a test to test batch insert case (though scala might not have it, we should have). You can refer to https://github.com/snowflakedb/snowpark-python/blob/ef0e8320feb374aab1949ce1b40fb69235fc4c24/tests/integ/test_dataframe.py#L1756-L1762